eth/db/header.py

```diff
@@ -146,46 +146,55 @@ def header_exists(self, block_hash: Hash32) -> bool:
         validate_word(block_hash, title="Block Hash")
         return block_hash in self.db
 
-    def persist_header(self,
-                       header: BlockHeader
-                       ) -> Tuple[Tuple[BlockHeader, ...], Tuple[BlockHeader, ...]]:
+    def persist_header_chain(self,
+                             *headers: BlockHeader
+                             ) -> Tuple[Tuple[BlockHeader, ...], Tuple[BlockHeader, ...]]:
         """
-        :returns: iterable of headers newly on the canonical chain
+        Return two iterable of headers, the first containing the new canonical headers,
+        the second containing the old canonical headers
         """
-        if header.parent_hash != GENESIS_PARENT_HASH:
-            try:
-                self.get_block_header_by_hash(header.parent_hash)
-            except HeaderNotFound:
+
+        previous = None
+
+        for header in headers:
+            is_genesis = header.parent_hash == GENESIS_PARENT_HASH
+            if not is_genesis and not self.header_exists(header.parent_hash):
                 raise ParentNotFound(
                     "Cannot persist block header ({}) with unknown parent ({})".format(
                         encode_hex(header.hash), encode_hex(header.parent_hash)))
 
-        self.db.set(
-            header.hash,
-            rlp.encode(header),
-        )
+            self.db.set(
+                header.hash,
+                rlp.encode(header),
+            )
 
-        if header.parent_hash == GENESIS_PARENT_HASH:
-            score = header.difficulty
-        else:
-            score = self.get_score(header.parent_hash) + header.difficulty
+            if is_genesis:
+                score = header.difficulty
+            elif previous is not None and previous[0].hash is header.parent_hash:
+                score = previous[1] + header.difficulty
+            else:
+                score = self.get_score(header.parent_hash) + header.difficulty
 
-        self.db.set(
-            SchemaV1.make_block_hash_to_score_lookup_key(header.hash),
-            rlp.encode(score, sedes=rlp.sedes.big_endian_int),
-        )
+            previous = (header, score)
+
+            self.db.set(
+                SchemaV1.make_block_hash_to_score_lookup_key(header.hash),
+                rlp.encode(score, sedes=rlp.sedes.big_endian_int),
+            )
 
         try:
             head_score = self.get_score(self.get_canonical_head().hash)
         except CanonicalHeadNotFound:
-            new_canonical_headers, old_canonical_headers = self._set_as_canonical_chain_head(
-                header.hash,
-            )
+            (
```

Comment:
In case someones wondering why the diff looks more messy then expected with respect to the change, this is because I moved the method from the `ChainDB` here which was just slightly nicer